drivers/bus/isapnp/stub.c

```diff
@@ -0,0 +1,263 @@
+/*
+ * PROJECT:     ReactOS ISA PnP Bus driver
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Stub driver
+ * COPYRIGHT:   Copyright 2021 Dmitry Borisov (di.sean@protonmail.com)
+ */
+
+/*
+ * This driver is only used if a platform has no ISA PnP support.
+ * We need to keep FDO because ACPI driver depends on it.
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <wdm.h>
+#include <section_attribs.h>
+
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS ********************************************************************/
+
+#define TAG_ISAPNP 'pasI'
+
+typedef struct _ISAPNP_STUB_EXTENSION
+{
+    PDEVICE_OBJECT Self;
+    PDEVICE_OBJECT Ldo;
+} ISAPNP_STUB_EXTENSION, *PISAPNP_STUB_EXTENSION;
+
+/* FUNCTIONS ******************************************************************/
+
+static DRIVER_DISPATCH_PAGED IsaStubCreateClose;
+
+static
+CODE_SEG("PAGE")
+NTSTATUS
+NTAPI
+IsaStubCreateClose(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _Inout_ PIRP Irp)
+{
+    PAGED_CODE();
+
+    Irp->IoStatus.Status = STATUS_SUCCESS;
+    IoCompleteRequest(Irp, IO_NO_INCREMENT);
+
+    return STATUS_SUCCESS;
+}
+
+static DRIVER_DISPATCH_PAGED IsaStubForward;
+
+static
+CODE_SEG("PAGE")
+NTSTATUS
+NTAPI
+IsaStubForward(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _Inout_ PIRP Irp)
+{
+    PISAPNP_STUB_EXTENSION FdoExt = DeviceObject->DeviceExtension;
+
+    PAGED_CODE();
+
+    IoSkipCurrentIrpStackLocation(Irp);
+    return IoCallDriver(FdoExt->Ldo, Irp);
+}
+
+static DRIVER_DISPATCH_PAGED IsaStubPnp;
+
+static
+CODE_SEG("PAGE")
+NTSTATUS
+NTAPI
+IsaStubPnp(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _Inout_ PIRP Irp)
+{
+    PISAPNP_STUB_EXTENSION FdoExt = DeviceObject->DeviceExtension;
+    PIO_STACK_LOCATION IrpSp = IoGetCurrentIrpStackLocation(Irp);
+    NTSTATUS Status;
+
+    PAGED_CODE();
+
+    DPRINT("%s(%p, %p) Minor - %X\n",
+           __FUNCTION__,
+           FdoExt,
+           Irp,
+           IrpSp->MinorFunction);
+
+    switch (IrpSp->MinorFunction)
+    {
+        case IRP_MN_START_DEVICE:
+        {
+            if (IoForwardIrpSynchronously(FdoExt->Ldo, Irp))
+                Status = Irp->IoStatus.Status;
+            else
+                Status = STATUS_UNSUCCESSFUL;
+
+            goto CompleteIrp;
+        }
+
+        case IRP_MN_QUERY_DEVICE_RELATIONS:
+        {
+            PDEVICE_RELATIONS DeviceRelations;
+
+            if (IrpSp->Parameters.QueryDeviceRelations.Type != BusRelations)
+                break;
+
+            DeviceRelations = ExAllocatePoolWithTag(PagedPool,
+                                                    FIELD_OFFSET(DEVICE_RELATIONS, Objects),
+                                                    TAG_ISAPNP);
```

Comment:
You seem to allocate this buffer but not initialize its contents , and you return it via the IRP.
This needs to be at least zeroed out: please use ExAllocatePoolZero().
That's because, even if you put the count to zero (below), you haven't put anything in its `Objects` field (whose first element is always present in the structure).
Or, put an explicit `DeviceRelations->Objects[0] = NULL;` below.

Reply:
I allocate space only for one `Count` member (via `FIELD_OFFSET`) so all contents will be initialized.

Reply:
It's fine for variable-length structures
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/io/pnpmgr/devaction.c;h=638dccb8dcd144fe70cfc288fc316a7b45d2e9f4;hb=HEAD#l2156
https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=+FIELD_OFFSET(DEVICE_